File: platform/platform-api/src/com/intellij/openapi/wm/ToolWindowManager.java
/*
 * Copyright 2000-2017 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */
package com.intellij.openapi.wm;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectManager;
import com.intellij.openapi.ui.MessageType;
import com.intellij.openapi.ui.popup.Balloon;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.HyperlinkListener;
import java.util.Objects;

/**
 * If you want to register a toolwindow, which will be enabled during the dumb mode, please use {@link ToolWindowManager}'s
 * registration methods which have 'canWorkInDumbMode' parameter.
 */
public abstract class ToolWindowManager {

  public abstract boolean canShowNotification(@NotNull String toolWindowId);

  public static ToolWindowManager getInstance(@NotNull Project project){
    return project.getComponent(ToolWindowManager.class);
  }

  /**
   * Register specified tool window into IDE window system.
   * @param id {@code id} of tool window to be registered.
   * @param component {@code component} which represents tool window content.
   * May be null. Content can be further added via content manager for this tool window (See {@link ToolWindow#getContentManager()})
   * @param anchor the default anchor for first registration. It uses only first time the
   * tool window with the specified {@code id} is being registered into the window system.
   * After the first registration window's anchor is stored in project file
   * and {@code anchor} is ignored.
   * @exception IllegalArgumentException if the same window is already installed or one
   * of the parameters is {@code null}.
   * @return tool window
   * @deprecated  {@link ToolWindowManager#registerToolWindow(String, boolean, ToolWindowAnchor)}
   */
  @Deprecated
  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id, @NotNull JComponent component, @NotNull ToolWindowAnchor anchor);

  /**
  * @deprecated  {@link ToolWindowManager#registerToolWindow(String, boolean, ToolWindowAnchor)}
  */
  @Deprecated
  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id,
                                                @NotNull JComponent component,
                                                @NotNull ToolWindowAnchor anchor,
                                                @NotNull Disposable parentDisposable);

  /**
  * @deprecated  {@link ToolWindowManager#registerToolWindow(String, boolean, ToolWindowAnchor)}
  */
  @Deprecated
  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id,
                                                @NotNull JComponent component,
                                                @NotNull ToolWindowAnchor anchor,
                                                @NotNull Disposable parentDisposable,
                                                boolean canWorkInDumbMode);
  /**
  * @deprecated  {@link ToolWindowManager#registerToolWindow(String, boolean, ToolWindowAnchor)}
  */
  @Deprecated
  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id,
                                                @NotNull JComponent component,
                                                @NotNull ToolWindowAnchor anchor,
                                                @NotNull Disposable parentDisposable,
                                                boolean canWorkInDumbMode,
                                                boolean canCloseContents);

  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id, boolean canCloseContent, @NotNull ToolWindowAnchor anchor);

  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id,
                                                boolean canCloseContent,
                                                @NotNull ToolWindowAnchor anchor,
                                                boolean secondary);

  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id,
                                                boolean canCloseContent,
                                                @NotNull ToolWindowAnchor anchor,
                                                @NotNull Disposable parentDisposable,
                                                boolean canWorkInDumbMode);

  @NotNull
  public abstract ToolWindow registerToolWindow(@NotNull String id,
                                                boolean canCloseContent,
                                                @NotNull ToolWindowAnchor anchor,
                                                @NotNull Disposable parentDisposable,
                                                boolean canWorkInDumbMode,
                                                boolean secondary);

  @NotNull
  public ToolWindow registerToolWindow(@NotNull final String id,
                                       final boolean canCloseContent,
                                       @NotNull final ToolWindowAnchor anchor,
                                       @NotNull Disposable parentDisposable) {
    return registerToolWindow(id, canCloseContent, anchor, parentDisposable, false);
  }

  /**
   * does nothing if tool window with specified isn't registered.
   */
  public abstract void unregisterToolWindow(@NotNull String id);

  /**
   */
  public abstract void activateEditorComponent();

  /**
   * @return {@code true} if and only if editor component is active.
   */
  public abstract boolean isEditorComponentActive();

  /**
   * @return array of {@code id}s of all registered tool windows.
   */
  @NotNull
  public abstract String[] getToolWindowIds();

  /**
   * @return {@code ID} of currently active tool window or {@code null} if there is no active
   * tool window.
   */
  @Nullable
  public abstract String getActiveToolWindowId();

  @Nullable
  public static ToolWindow getActiveToolWindow () {
    IdeFrame frame = IdeFocusManager.getGlobalInstance().getLastFocusedFrame();
    Project project = frame == null ? ProjectManager.getInstance().getDefaultProject() : frame.getProject();

    if (project != null) {
      ToolWindowManager managerInstance = getInstance(project);
      if (managerInstance != null) {
        return managerInstance.getToolWindow(getActiveId());
      }
    }

    return null;
  }

  @Nullable
  public static String getActiveId () {
    IdeFrame lastFocusedFrame = IdeFocusManager.getGlobalInstance().getLastFocusedFrame();

    if (lastFocusedFrame !=null) {
      Project project = lastFocusedFrame.getProject();
      if (project != null) {
        ToolWindowManager instance = getInstance(project);
        return instance == null ? "" : instance.getActiveToolWindowId();
      }
    }
    return "";
  }


  /**
   * @return registered tool window with specified {@code id}. If there is no registered
   * tool window with specified {@code id} then the method returns {@code null}.
   */
  public abstract ToolWindow getToolWindow(@Nullable String id);

  /**
   * Puts specified runnable to the tail of current command queue.
   */
  public abstract void invokeLater(@NotNull Runnable runnable);

  /**
   * Utility method for quick access to the focus manager
   */
  @NotNull
  public abstract IdeFocusManager getFocusManager();

  public abstract void notifyByBalloon(@NotNull final String toolWindowId, @NotNull final MessageType type, @NotNull final String htmlBody);

  public abstract void notifyByBalloon(@NotNull final String toolWindowId,
                                       @NotNull final MessageType type,
                                       @NotNull final String htmlBody,
                                       @Nullable final Icon icon,
                                       @Nullable HyperlinkListener listener);

  @Nullable
  public abstract Balloon getToolWindowBalloon(String id);

  public abstract boolean isMaximized(@NotNull ToolWindow wnd);

  public abstract void setMaximized(@NotNull ToolWindow wnd, boolean maximized);
}

File: platform/platform-api/src/com/intellij/openapi/wm/ToolWindowManager.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.openapi.wm

import com.intellij.openapi.Disposable
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.MessageType
import com.intellij.openapi.ui.popup.Balloon
import com.intellij.openapi.ui.popup.BalloonBuilder
import com.intellij.openapi.util.NlsContexts
import org.jetbrains.annotations.ApiStatus
import org.jetbrains.annotations.ApiStatus.Internal
import org.jetbrains.annotations.NonNls
import java.util.function.Consumer
import java.util.function.Supplier
import javax.swing.Icon
import javax.swing.JComponent
import javax.swing.event.HyperlinkListener

/**
 * If you want to register a toolwindow, which will be enabled during the dumb mode, please use [ToolWindowManager]'s
 * registration methods which have 'canWorkInDumbMode' parameter.
 *
 * @see com.intellij.openapi.wm.ex.ToolWindowManagerListener
 */
abstract class ToolWindowManager {
  companion object {
    @JvmStatic
    fun getInstance(project: Project): ToolWindowManager = project.service<ToolWindowManager>()
  }

  abstract val focusManager: IdeFocusManager

  abstract fun canShowNotification(toolWindowId: String): Boolean

  @Suppress("DeprecatedCallableAddReplaceWith")
  @Deprecated("Use ToolWindowFactory and com.intellij.toolWindow extension point")
  fun registerToolWindow(id: String, component: JComponent, anchor: ToolWindowAnchor): ToolWindow {
    return registerToolWindow(RegisterToolWindowTask(id = id, component = component, anchor = anchor, canCloseContent = false, canWorkInDumbMode = false))
  }

  @Suppress("DeprecatedCallableAddReplaceWith")
  @Deprecated("Use ToolWindowFactory and com.intellij.toolWindow extension point")
  fun registerToolWindow(id: String, canCloseContent: Boolean, anchor: ToolWindowAnchor): ToolWindow {
    return registerToolWindow(RegisterToolWindowTask(id = id, anchor = anchor, canCloseContent = canCloseContent, canWorkInDumbMode = false))
  }

  @Suppress("DeprecatedCallableAddReplaceWith")
  @Deprecated("Use ToolWindowFactory and com.intellij.toolWindow extension point")
  fun registerToolWindow(id: String,
                         canCloseContent: Boolean,
                         anchor: ToolWindowAnchor,
                         secondary: Boolean): ToolWindow {
    return registerToolWindow(RegisterToolWindowTask(id = id, anchor = anchor, sideTool = secondary, canCloseContent = canCloseContent, canWorkInDumbMode = false))
  }

  @Suppress("DeprecatedCallableAddReplaceWith")
  @Deprecated("Use ToolWindowFactory and com.intellij.toolWindow extension point")
  fun registerToolWindow(id: String,
                         canCloseContent: Boolean,
                         anchor: ToolWindowAnchor,
                         @Suppress("UNUSED_PARAMETER") parentDisposable: Disposable,
                         canWorkInDumbMode: Boolean): ToolWindow {
    return registerToolWindow(RegisterToolWindowTask(id = id, anchor = anchor, canCloseContent = canCloseContent, canWorkInDumbMode = canWorkInDumbMode))
  }

  @Suppress("DeprecatedCallableAddReplaceWith")
  @Deprecated("Use ToolWindowFactory and com.intellij.toolWindow extension point")
  fun registerToolWindow(id: String,
                         canCloseContent: Boolean,
                         anchor: ToolWindowAnchor,
                         @Suppress("UNUSED_PARAMETER") parentDisposable: Disposable,
                         canWorkInDumbMode: Boolean,
                         secondary: Boolean): ToolWindow {
    return registerToolWindow(RegisterToolWindowTask(id = id, anchor = anchor, sideTool = secondary, canCloseContent = canCloseContent, canWorkInDumbMode = canWorkInDumbMode))
  }

  @Suppress("DeprecatedCallableAddReplaceWith")
  @Deprecated("Use ToolWindowFactory and com.intellij.toolWindow extension point")
  @ApiStatus.ScheduledForRemoval
  fun registerToolWindow(id: String,
                         canCloseContent: Boolean,
                         anchor: ToolWindowAnchor,
                         @Suppress("UNUSED_PARAMETER") parentDisposable: Disposable): ToolWindow {
    return registerToolWindow(RegisterToolWindowTask(id = id, anchor = anchor, canCloseContent = canCloseContent, canWorkInDumbMode = false))
  }

  @Internal
  abstract fun registerToolWindow(task: RegisterToolWindowTask): ToolWindow

  /**
   * Use only for dynamically registered toolwindows required for a certain operation.
   *
   * [ToolWindow.getAnchor] is set to [ToolWindowAnchor.BOTTOM] by default.
   * [ToolWindow.setToHideOnEmptyContent] is set to `true` by default.
   */
  inline fun registerToolWindow(id: String, builder: RegisterToolWindowTaskBuilder.() -> Unit): ToolWindow {
    val b = RegisterToolWindowTaskBuilder(id)
    b.builder()
    return registerToolWindow(b.build())
  }

  /**
   * Does nothing if a tool window with specified id isn't registered.
   */
  @Deprecated("Use ToolWindowFactory and com.intellij.toolWindow extension point")
  abstract fun unregisterToolWindow(id: String)

  abstract fun activateEditorComponent()

  /**
   * @return `true` if and only if an editor component is active.
   */
  abstract val isEditorComponentActive: Boolean

  /**
   * @return array of `id`s of all registered tool windows.
   */
  abstract val toolWindowIds: Array<String>

  abstract val toolWindowIdSet: Set<String>

  /**
   * @return `ID` of currently active tool window or `null` if there is no active tool window.
   */
  abstract val activeToolWindowId: String?

  /**
   * @return `ID` of tool window activated last time.
   */
  abstract val lastActiveToolWindowId: String?

  /**
   * @return registered tool window with specified `id`. If there is no registered
   * tool window with specified `id` then the method returns `null`.
   * @see ToolWindowId
   */
  abstract fun getToolWindow(@NonNls id: String?): ToolWindow?

  /**
   * Puts specified runnable to the tail of the current command queue.
   */
  abstract fun invokeLater(runnable: Runnable)

  fun notifyByBalloon(toolWindowId: String, type: MessageType, @NlsContexts.NotificationContent htmlBody: String) {
    notifyByBalloon(ToolWindowBalloonShowOptions(toolWindowId, type, htmlBody))
  }

  fun notifyByBalloon(toolWindowId: String,
                      type: MessageType,
                      @NlsContexts.PopupContent htmlBody: String,
                      icon: Icon?,
                      listener: HyperlinkListener?) {
    notifyByBalloon(ToolWindowBalloonShowOptions(toolWindowId, type, htmlBody, icon, listener))
  }

  abstract fun notifyByBalloon(options: ToolWindowBalloonShowOptions)

  abstract fun getToolWindowBalloon(id: String): Balloon?

  @Internal
  open fun closeBalloons() {
  }

  abstract fun isMaximized(window: ToolWindow): Boolean

  abstract fun setMaximized(window: ToolWindow, maximized: Boolean)

  /*
   * Returns visual representation of tool window location
   * @see AllIcons.Actions#MoveToBottomLeft ... com.intellij.icons.AllIcons.Actions#MoveToWindow icon set
   */
  open fun getLocationIcon(id: String, fallbackIcon: Icon): Icon = fallbackIcon

  open fun isStripeButtonShow(toolWindow: ToolWindow): Boolean = false
}

class RegisterToolWindowTaskBuilder @PublishedApi internal constructor(private val id: String) {
  @JvmField
  var anchor: ToolWindowAnchor = ToolWindowAnchor.BOTTOM
  @JvmField
  var stripeTitle: Supplier<@NlsContexts.TabTitle String>? = null
  @JvmField
  var icon: Icon? = null
  @JvmField
  var shouldBeAvailable: Boolean = true
  @JvmField
  var canCloseContent: Boolean = true
  @JvmField
  var hideOnEmptyContent: Boolean = true
  @JvmField
  var sideTool: Boolean = false

  @JvmField
  var contentFactory: ToolWindowFactory? = null

  @PublishedApi
  internal fun build(): RegisterToolWindowTask {
    val result = RegisterToolWindowTask(id = id,
                                        anchor = anchor,
                                        component = null,
                                        sideTool = sideTool,
                                        canCloseContent = canCloseContent,
                                        shouldBeAvailable = shouldBeAvailable,
                                        contentFactory = contentFactory,
                                        icon = icon,
                                        stripeTitle = stripeTitle)

    result.hideOnEmptyContent = hideOnEmptyContent
    return result
  }
}

data class ToolWindowBalloonShowOptions(val toolWindowId: String,
                                        val type: MessageType,
                                        @NlsContexts.PopupContent val htmlBody: String,
                                        val icon: Icon? = null,
                                        val listener: HyperlinkListener? = null,
                                        val balloonCustomizer: Consumer<BalloonBuilder>? = null)